_user/rhealstone/deadlock-break/main.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Rhealstone benchmark
+ *
+ * Task switch
+ *
+ * Copyright 2024 Phoenix Systems
+ * Author: Lukasz Leczkowski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdatomic.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/time.h>
+#include <sys/threads.h>
+
+#include <board_config.h>
+#include <rhcommon.h>
+
+
+#define BENCHMARKS 1000
+
+#define ONE_TICK     26000 /* Number dependent on CPU. For loop to this count must be longer than sleep period (1 ms) */
+#define ONE_TICK_AVG 24990 /* For loop up to this count takes 1ms */
+
+bool deadBrk;
+atomic_int count = 0;
+atomic_bool done = false;
+handle_t mutex;
+
+unsigned char stack[3][4096];
+volatile atomic_int i;
+int maxLoops = BENCHMARKS;
+int tid1, tid2, tid3;
+
+atomic_bool t3_started = false;
+
+/* High priority task */
+void task3(void *arg)
+{
+	t3_started = true;
+	if (deadBrk) {
+		mutexLock(mutex);
+	}
+
+	if (deadBrk) {
+		mutexUnlock(mutex);
+	}
+	done = true;
+	endthread();
+}
+
+
+/* Medium priority task */
+void task2(void *arg)
+{
+	if (beginthreadex(task3, 1, stack[2], sizeof(stack[2]), NULL, &tid3)) {
+		perror("beginthreadex");
+		endthread();
+	}
+
+	while (!done) {
+		__asm__ volatile("nop");
+	}
+
+	threadJoin(tid3, 0);
+
+	endthread();
+}
+
+/* Low priority task */
+void task1(void *arg)
+{
+	if (deadBrk) {
+		mutexLock(mutex);
+	}
+
+	if (beginthreadex(task2, 2, stack[1], sizeof(stack[1]), NULL, &tid2)) {
+		perror("beginthreadex");
+		endthread();
+	}
+
+	while (!t3_started) {
+		usleep(0);
+	}
+
+	if (deadBrk) {
+		mutexUnlock(mutex);
+	}
+
+	threadJoin(tid2, 0);
+	t3_started = false;
+	endthread();
+}
+
+
+uint64_t totalNoDeadBrk = 0;
+uint64_t totalDeadBrk = 0;
+
+
+int doTest(void)
+{
+	if (beginthreadex(task1, 3, stack[0], sizeof(stack[0]), NULL, &tid1)) {
+		perror("beginthreadex");
+		return -1;
+	}
+
+	uint64_t benchStart, benchEnd;
+
+	benchStart = getCntr();
+
+	priority(4);
+
+	usleep(0);
+
+	threadJoin(tid1, 0);
+
+	benchEnd = getCntr();
```

Comment:
We introduce 2 syscalls overhead here that skews the result. Perhaps we could measure the time in `task1` instead?